src/lang/eval/StateService.ml

```diff
@@ -179,7 +182,12 @@ let update_local ~fname ~keys vopt fields =
 let update ~fname ~keys ~value =
   let%bind (sm, fields) = assert_init() in
   match sm with
-  | IPC _ -> fail0 "StateService: IPC state service is unimplemented"
+  | IPC socket_addr ->
+    let%bind tp = field_type fields fname in
+    let%bind _ = StateIPCClient.update ~socket_addr ~fname ~keys ~value ~tp in
+    if keys <> []
+      then pure @@ (G_MapUpdate(List.length keys, Some value))
+      else pure @@ (G_Store(value))
   | Local ->
     let%bind (fields', g) = update_local ~fname ~keys (Some value) fields in
```

Comment:
There is no change in gas accounting with IPC. In the old-style, they are done within the actual implementation functions (`fetch_local`), but in the new style, instead of having `StateIPCClient.ml` deal with gas, I've retained them in `StateService.ml`, which is nearer to Eval.

Gas costs are _identical_ in IPC and non-IPC modes.